File: leetcode-in-python/solution/7_reverse_integer.py
class Solution:
    def reverse(self, x: int) -> int:
        ans = 0
        # [-2147483648, 2147483647]
        MAX_INT = 2 ** 31 - 1
        MIN_INT = -2 ** 31

        while x != 0:
            if x > 0:
                pop = x % 10
                x = x // 10
            else:
                pop = x % - 10
                x = -(x // -10)

            if ans > MAX_INT / 10 or (ans == MAX_INT and pop > 7):  # 2147483647 = 2147483640 + 7
                return 0
            if ans < MIN_INT / 10 or (ans == MIN_INT and pop < -8):  # -2147483648 = -2147483640 +(-8)
                return 0
            ans = ans * 10 + pop
        return ans

import unittest


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.input = [123, 0, -123, 2147483647, 1463847412, 1463847413, 1534236469]
        self.answer = [321, 0, -321, 0, 2147483641, 0, 0]
        self.s = Solution()

    def test_solution(self):
        for i, item in enumerate(self.input):
            ans = self.s.reverse(item)
            self.assertEqual(ans , self.answer[i],ans )


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_7_reverse_integer.rs
pub struct Solution;

impl Solution {
    pub fn reverse(x: i32) -> i32 {
        Solution::_reverse(x).unwrap_or_default()
    }

    fn _reverse(mut x: i32) -> Option<i32> {
        let mut ans = 0i32;
        while x != 0 {
            let pop = x % 10;
            ans = ans.checked_mul(10)?.checked_add(pop)?;
            x /= 10;
        }
        Some(ans)
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (123, 321),
            (0, 0),
            (-123, -321),
            (2147483647, 0),
            (1463847412, 2147483641),
            (1463847413, 0),
            (1534236469, 0)
        ];

        for (x, answer) in test_case {
            let ans = Solution::reverse(x);
            assert_eq!(answer, ans);
        }
    }
}
